pkg/receive/handler.go

```diff
@@ -0,0 +1,114 @@
+package receive
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"strings"
+	"time"
+)
+
+type clusterIDKey string
+
+const forwardTimeout = 5 * time.Second
+const clusterID clusterIDKey = "tenant"
+
+// ClusterAuthorizer authorizes a cluster by its token and id, returning a subject or error
+type ClusterAuthorizer interface {
+	AuthorizeCluster(token, cluster string) (subject string, err error)
+}
+
+// Handler knows the forwardURL for all requests
+type Handler struct {
+	ForwardURL string
+	client     *http.Client
+}
+
+// NewHandler returns a new Handler with a http client
+func NewHandler(forwardURL string) *Handler {
+	return &Handler{
+		ForwardURL: forwardURL,
+		client: &http.Client{
+			Timeout: forwardTimeout,
+		},
+	}
+}
+
+// Receive a remote-write request after it has been authenticated and forward it to Thanos
+func (h *Handler) Receive(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	defer r.Body.Close()
+
+	ctx, cancel := context.WithTimeout(r.Context(), forwardTimeout)
+	defer cancel()
+
+	req, err := http.NewRequest(http.MethodPost, h.ForwardURL, r.Body)
+	if err != nil {
+		log.Printf("failed to create forward request: %v\n", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	req = req.WithContext(ctx)
+	req.Header.Add("THANOS-TENANT", r.Context().Value(clusterID).(string))
+
+	resp, err := h.client.Do(req)
+	if err != nil {
+		log.Printf("failed to foward request: %v\n", err)
+		http.Error(w, err.Error(), http.StatusBadGateway)
+		return
+	}
+
+	if resp.StatusCode/100 != 2 {
+		log.Printf("response status code is %s\n", resp.Status)
+		http.Error(w, "upstream response status is not 200 OK", http.StatusBadGateway)
+		return
+	}
+}
+
+type AuthorizationPayload struct {
+	Cluster string `json:"cluster"`
```

Comment:
This'll work for now, but doesn't generally work. We need to rethink this a bit. Not a blocker on this though.

Reply:
Ack, this is 100% specific to Tollbooth. We’ll need to figure out a way to generalize this and maintain compat with Tollbooth auth

Reply:
I just realized that there is a decent way we can keep this generic and not tightly coupled to tollbooth. tollbooth expects all authorization requests to be POST requests with the following body:
```json
{"cluster_id":"<cluster_id>","authorization_token":"<pull_secret_auth_token>"}
```
We can ensure that the bearer token that prometheus is configured to send matches that format and then we POST that token to the specified URL. This way we don't have to do any custom marshalling/unmarshalling. The only variables then are the authorization URL and the token contents.

The one thing that is funny is that we are taking the bearer token header and using that as the body to of a POST request, but I still think that's better and more generic than what we have today.